src/Google/GlobalSiteTag.php

```diff
@@ -156,14 +195,241 @@ public function maybe_display_event_snippet( string $ads_conversion_id, string $
+			$total_orders = wc_get_customer_order_count( $order->get_user_id() );
+		} else {
+			$total_orders = WC_Order_Export_Data_Extractor::get_customer_order_count_by_email( $order->get_billing_email() );
+		}
+		$is_new_customer = ( $total_orders === 1 ) ? 'true' : 'false';
+		$language        = $this->wp->get_locale();
+		if ( 'en_US' === $language ) {
+			$language = 'English';
+		}
+		printf(
+			'<script>
+            gtag("event", "purchase",
+				{
+                    "developer_id.%s": "true",
+                    "ecomm_pagetype": "purchase",
+                    "send_to": "GLA",
+                    transaction_id": "%s",
+                    "currency": "%s",
+                    "country": "%s",
+                    "value": "%s",
+                    "new_customer": "%s",
+                    "tax": "%s",
+                    "shipping": "%s",
+                    "delivery_posatal_code": "%s",
+                    "aw_feed_country": "%s",   
+                    "aw_feed_language": "%s",                 
+                    items: [' . esc_html( $item_info ) . ']}); </script>',
+			esc_js( self::DEVELOPER_ID ),
+			esc_js( $order->get_id() ),
+			esc_js( $order->get_currency() ),
+			esc_js( WC()->countries->get_base_country() ),
+			esc_js( $order->get_total() ),
+			esc_js( $is_new_customer ),
+			esc_js( $order->get_cart_tax() ),
+			esc_js( $order->get_total_shipping() ),
+			esc_js( $order->get_shipping_postcode() ),
+			esc_js( WC()->countries->get_base_country() ),
+			esc_js( $language ),
+		);
+	}
+
+	/**
+	 * Display the JavaScript code to track the add to cart button.
+	 *
+	 * @param string $message Add to cart messages.
+	 * @param array  $products Product ID list.
+	 */
+	public function custom_action_add_to_cart( $message, $products ) {
+		// Only display this tag info after click the add to cart button .
+		$product = wc_get_product( array_key_first( $products ) );
+
+		add_action(
+			'wp_body_open',
+			function () use ( $product ) {
+				printf(
+					'<script>
+                        gtag("event", "add_to_cart", {
+                            "send_to": "GLA",
+                            "developer_id.%s": "true",
+                            "ecomm_pagetype": "cart",
+                            "value": "%s",
+                            items:[{
+                            "id": "gla_%s",
+                            "price": %s,
+                            "google_business_vertical": "retail",
+                            "name":"%s",
+                            "category":"%s",
+                            }]});
+                    </script>',
+					esc_js( self::DEVELOPER_ID ),
+					esc_js( (string) $product->get_price() ),
+					esc_js( $product->get_id() ),
+					esc_js( (string) $product->get_price() ),
+					esc_js( $product->get_name() ),
+					esc_js( join( '& ', $product->get_categories() ) ),
+				);
+			},
+			1000005
+		);
+
+		do_action(
+			'wp_body_open'
+		);
```

Comment:
This is not the right place to call this action, as this causes the action to be called twice, and any function that hooks into this action will be called again here. If we want this JS to be output after this filter is called, it's probably better to hook it to `wp_footer`. Although we are probably going to need to adjust this anyways to make it compatible with blocks.

Reply:
Thanks for your suggestions. Updated.